### PR TITLE
Updated Operator SDK to v0.18.2

### LINKED ISSUE #1125: Upgrade Operator SDK to v0.18.2

### PR DESCRIPTION
**Pull request by jpkrohling**
Fixes #1125

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

File: pkg/ingress/ingress_test.go [hunk 1]
```diff
@@ -89,6 +87,8 @@ func TestIngressNetworkingAPI(t *testing.T) {
 }
 
 func TestIngressExtensionAPI(t *testing.T) {
+	t.Skip("test skipped after migrating to Operator SDK 1.18.2")
```
Comment (jpkrohling) [inline pkg/ingress/ingress_test.go:90]: @rubenvp8510 I wasn't able to understand what this test is actually trying to accomplish. From what I understood, there is (was?) a conversion between a legacy ingress type into a newer type. Could you please confirm that this is still happening? Or perhaps you can spot what I did wrong in this test, causing it to fail...

**Comment by jpkrohling** [comment]
Apparently, the coverage change is related to the skipping of the ingress test.

File: Makefile [hunk 2]
```diff
@@ -259,7 +259,7 @@ kafka: deploy-kafka-operator
 	@kubectl create namespace $(KAFKA_NAMESPACE) 2>&1 | grep -v "already exists" || true
 	@curl --fail --location $(KAFKA_EXAMPLE) --output deploy/test/kafka-example.yaml --create-dirs
 	@sed -i 's/size: 100Gi/size: 10Gi/g' deploy/test/kafka-example.yaml
-	@kubectl -n $(KAFKA_NAMESPACE) apply --dry-run=client -f deploy/test/kafka-example.yaml
+	@kubectl -n $(KAFKA_NAMESPACE) apply --dry-run=true -f deploy/test/kafka-example.yaml
```
Comment (jpkrohling) [inline Makefile:262]: @kevinearls I think you added this flag in the past. Not sure where this option is parsed, but it currently fails with my client/server combination. I got it working by setting it to `client`, but wanted to check with you whether this is indeed the correct value.

For reference, here are the versions I'm using:

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0-4-g38212b5", GitCommit:"d038424d6d4f1cc39ad586ac0d36dac3a7a37ceb", GitTreeState:"clean", BuildDate:"2020-06-16T03:29:46Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

File: deploy/cluster_role.yaml [hunk 3]
```diff
@@ -15,12 +15,13 @@ rules:
   resources:
   - '*'
   verbs:
-  - 'get'
-  - 'list'
-  - 'create'
-  - 'update'
-  - 'delete'
-  - 'watch'
+  - create
```
Comment (jpkrohling) [inline deploy/cluster_role.yaml:18]: This change is mainly to reduce the diff between new operators and this operator, in an attempt to make it easier to compare what changed between SDK versions.

The new list contains `patch`, which the old list didn't. Apart from that, it's only ordering and formatting changes.

File: test/e2e/all_in_one_test.go [hunk 4]
```diff
@@ -41,8 +41,8 @@ func (suite *AllInOneTestSuite) SetupSuite() {
 		require.FailNow(t, "Failed in prepare")
 	}
 	fw = framework.Global
-	namespace, _ = ctx.GetNamespace()
```
Comment (jpkrohling) [inline test/e2e/all_in_one_test.go:44]: @kevinearls heads up: `GetNamespace` actually creates a new namespace (!!), which fails during runtime as the namespace already exists. Behind the scenes, `GetNamespace` calls `GetID` to obtain the namespace's value, hence this change.

File: test/e2e/utils.go [hunk 5]
```diff
@@ -14,16 +14,13 @@ import (
...
-	"github.com/jaegertracing/jaeger-operator/pkg/util"
-
 	osv1 "github.com/openshift/api/route/v1"
 	osv1sec "github.com/openshift/api/security/v1"
 	framework "github.com/operator-framework/operator-sdk/pkg/test"
 	"github.com/operator-framework/operator-sdk/pkg/test/e2eutil"
 	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"golang.org/x/net/context"
```
Comment (jpkrohling) [inline test/e2e/utils.go:26]: This seems to be added via 5c37d2df44ee7f0017103bf2366b2caac897939e , but I guess this was supposed to be `context` instead. Just wanted to confirm with you if you really meant this old library, or if we can safely change it to use the stdlib's `context`.

File: pkg/ingress/ingress_test.go (hunk 1 above)
Comment (rubenvp8510) [reply pkg/ingress/ingress_test.go:90]: The conversion was the intention of this PR.  Is this test failing with the SDK update? This was introduced for maintain backward compatibility as an step to migrate to the newer networking.k8s.io/v1beta1 API.

Comment (rubenvp8510) [reply pkg/ingress/ingress_test.go:90]: The problem is: at the time I did the test I created an object `Ingress` with the `namespace` and `name` that match the object I want to update, then send it as an update to the API. That is an incorrect approach, but for some reason it wasn't complain.

Now you fixed it, the problem with your approach is that you are using the new `Network` API to get the ingress resource, but it doesn't exist because it was created by the old `extension` API. Indeed is what this test is testing that we are using the old API if the `ingress-api` flag is set to use it.


I'll change the following lines 
```

updatedIngress := &netv1beta.Ingress{}
	err = cl.Get(context.Background(), types.NamespacedName{Name: ingress.GetName(), Namespace: ingress.GetNamespace()}, updatedIngress)
	require.NoError(t, err)


```
With this:

```
	updatedIngressExt := &extv1beta.Ingress{}

	err = cl.Get(context.Background(), types.NamespacedName{Name: ingress.GetName(), Namespace: ingress.GetNamespace()}, updatedIngressExt)
	require.NoError(t, err)
	updatedIngress := ingressClient.fromExtToNet(*updatedIngressExt)

	updatedIngress.Spec.Backend.ServiceName = "updated-srv"
	err = ingressClient.Update(context.Background(), &updatedIngress)
	require.NoError(t, err)
```

Said that, you haven't broken anything :) So if you want I can fix the test in a following PR.

Comment (jpkrohling) [reply pkg/ingress/ingress_test.go:90]: It now passes, thanks!

**Comment by kevinearls** [comment]
@jpkrohling I am getting a failure running E2E tests on an OCP cluster, although strangely it's not 100% clear what is failing or why.  I will work on this more later today.

File: deploy/cluster_role.yaml [hunk 6]
```diff
@@ -181,21 +192,22 @@ rules:
   resources:
   - deployments
   verbs:
-  - 'get'
-  - 'list'
-  - 'create'
```
Comment (objectiser) [inline deploy/cluster_role.yaml:186]: Did you mean to drop `create` here?

File: deploy/olm-catalog/jaeger-operator/1.18.1/jaeger-operator.v1.18.1.clusterserviceversion.yaml [hunk 7]
```diff
@@ -279,19 +290,20 @@ spec:
           verbs:
           - get
           - list
-          - create
```
Comment (objectiser) [inline deploy/olm-catalog/jaeger-operator/1.18.1/jaeger-operator.v1.18.1.clusterserviceversion.yaml:282]: And here `create` is dropped?

File: pkg/apis/jaegertracing/v1/jaeger_types.go [hunk 8]
```diff
@@ -153,11 +153,11 @@ type Jaeger struct {
 // +k8s:openapi-gen=true
 type JaegerCommonSpec struct {
 	// +optional
-	// +listType=set
+	// +listType=atomic
```
Comment (objectiser) [inline pkg/apis/jaegertracing/v1/jaeger_types.go:156]: Why these changes?

File: go.mod [hunk 9]
```diff
@@ -4,58 +4,33 @@ go 1.13
 
 require (
```
Comment (pavolloffay) [inline go.mod:5]: Update also the go version to 1.14

File: go.mod [hunk 10]
```diff
@@ -4,58 +4,33 @@ go 1.13
...
+	google.golang.org/grpc v1.30.0
+	k8s.io/api v0.18.3
+	k8s.io/apimachinery v0.18.3
 	k8s.io/client-go v12.0.0+incompatible
-	k8s.io/kube-openapi v0.0.0-20200204173128-addea2498afe
-	sigs.k8s.io/controller-runtime v0.4.0
+	k8s.io/kube-openapi v0.0.0-20200410145947-61e04a5be9a6
+	sigs.k8s.io/controller-runtime v0.6.0
+
+	// testing
```
Comment (pavolloffay) [inline go.mod:25]: isn't this removed with `go mod tidy`?

File: pkg/autodetect/main.go [hunk 11]
```diff
@@ -197,6 +197,7 @@ func (b *Background) detectKafka(ctx context.Context, apiList *metav1.APIGroupLi
 
 func (b *Background) detectClusterRoles(ctx context.Context) {
 	tr := &authenticationapi.TokenReview{
+		ObjectMeta: metav1.ObjectMeta{Name: "jaeger-operator-TEST"},
```
Comment (pavolloffay) [inline pkg/autodetect/main.go:200]: leftover?

File: pkg/cmd/start/main.go [hunk 12]
```diff
@@ -108,6 +109,10 @@ func NewStartCommand() *cobra.Command {
 
 func start(cmd *cobra.Command, args []string) {
 	mgr := bootstrap(context.Background())
+
 	// Start the Cmd
-	log.Fatal(mgr.Start(signals.SetupSignalHandler()))
+	if err := mgr.Start(signals.SetupSignalHandler()); err != nil {
+		log.Fatal(err, "Manager exited non-zero")
+		os.Exit(1)
```
Comment (pavolloffay) [inline pkg/cmd/start/main.go:116]: the line above should already exit with 1

File: pkg/service/collector.go [hunk 13]
```diff
@@ -7,9 +7,10 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/spf13/viper"
```
Comment (pavolloffay) [inline pkg/service/collector.go:10]: It should be moved to group 2

**Comment by kevinearls** [comment]
I found a few problems with tests:

W0710 09:17:42.229260   17185 helpers.go:549] --dry-run=true is deprecated (boolean value) and can be replaced with --dry-run=client.
# github.com/jaegertracing/jaeger-operator/test/e2e [github.com/jaegertracing/jaeger-operator/test/e2e.test]
test/e2e/self_provisioned_elasticsearch_test.go:143:72: not enough arguments in call to fw.KubeClient.AppsV1().Deployments(namespace).List
	have ("k8s.io/apimachinery/pkg/apis/meta/v1".ListOptions)
	want (context.Context, "k8s.io/apimachinery/pkg/apis/meta/v1".ListOptions)
test/e2e/self_provisioned_elasticsearch_test.go:162:59: not enough arguments in call to fw.KubeClient.CoreV1().Pods(namespace).List
	have ("k8s.io/apimachinery/pkg/apis/meta/v1".ListOptions)
	want (context.Context, "k8s.io/apimachinery/pkg/apis/meta/v1".ListOptions)
test/e2e/self_provisioned_elasticsearch_test.go:258:67: not enough arguments in call to kubeclient.AppsV1().Deployments(namespace).Get
	have (string, "k8s.io/apimachinery/pkg/apis/meta/v1".GetOptions)
	want (context.Context, string, "k8s.io/apimachinery/pkg/apis/meta/v1".GetOptions)
make: *** [e2e-tests-self-provisioned-es] Error 2

File: go.mod (hunk 9 above)
Comment (jpkrohling) [reply go.mod:5]: I thought about having the CI running at 1.14 first, and then updating this. I'm not sure our downstream builds are able to run 1.14 yet.

cc @kevinearls and @objectiser

File: go.mod (hunk 10 above)
Comment (jpkrohling) [reply go.mod]: It is :/ I just changed the PR to remove this, and applied a `go mod tidy`.

File: pkg/autodetect/main.go (hunk 11 above)
Comment (jpkrohling) [reply pkg/autodetect/main.go:200]: No, this is actually required now, otherwise it fails to validate ("token review has to have a name" or something...)

File: go.mod (hunk 9 above)
Comment (objectiser) [reply go.mod:5]: Correct - still on 1.13 - however we should be able to move to 1.14 here and pick up any issues when the internal CI builds run.

File: pkg/apis/jaegertracing/v1/jaeger_types.go (hunk 8 above)
Comment (jpkrohling) [reply pkg/apis/jaegertracing/v1/jaeger_types.go:156]: There was a change in how OpenAPI and the generated resources would validate the values. I don't remember exactly the reason, as this is something I faced during the upgrade of the OpenTelemetry Operator as well: https://github.com/open-telemetry/opentelemetry-operator/pull/24/files#diff-0f1140984003d7bcbaddb3717895a48d

File: deploy/cluster_role.yaml (hunk 6 above)
Comment (jpkrohling) [reply deploy/cluster_role.yaml:186]: Yes, these are for the `inject` feature, so, we just want to change existing deployments, not create new ones. But you made me realize that this doesn't make sense, as we have full right for `deployments` defined a few lines above... 

Created #1130 to track this.

**Comment by jpkrohling** [comment]
@pavolloffay said:
> Does this PR pass the full e2e tests including e2e-tests-self-provisioned-es on OCP?

I think those are the ones that @kevinearls mentioned a few minutes ago and fixed in his branch. I cherry-picked his commit and added to this PR. 

I also addressed the other comments. The only open comment is about the change from `set` to `atomic` (@objectiser). If more details about is needed, I can revert that and trace back why it's needed.

**Comment by kevinearls** [comment]
@jpkrohling @pavolloffay I did a rerun on an OCP 4.4 cluster and all tests have passed, including the self provisioned es tests.

File: pkg/autodetect/main.go (hunk 11 above)
Comment (pavolloffay) [reply pkg/autodetect/main.go:200]: Could this cause any issues when with multiple operands in the same namespace or multiple operators in the cluster?

File: pkg/apis/jaegertracing/v1/jaeger_types.go (hunk 8 above)
Comment (objectiser) [reply pkg/apis/jaegertracing/v1/jaeger_types.go:156]: ok thanks

File: test/e2e/all_in_one_test.go [hunk 14]
```diff
@@ -41,8 +41,8 @@ func (suite *AllInOneTestSuite) SetupSuite() {
...
-	require.NotNil(t, namespace, "GetNamespace failed")
+	namespace = ctx.GetID()
+	require.NotNil(t, namespace, "GetID failed")
```
Comment (chlunde) [inline test/e2e/all_in_one_test.go:45]: Should these be NotEmpty now? But I'm not sure it can fail at this point

File: Makefile [hunk 15]
```diff
@@ -65,6 +65,8 @@ security:
 build: format
```
Comment (chlunde) [inline Makefile:65]: `SDK_VERSION` is not updated above in this file

Comment (chlunde) [reply Makefile:65]: https://github.com/jaegertracing/jaeger-operator/pull/1133
